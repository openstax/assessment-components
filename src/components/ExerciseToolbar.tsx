import { colors, layouts, breakpoints } from "../../src/theme";
import styled, { css } from "styled-components";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBookOpen, faTriangleExclamation } from "@fortawesome/free-solid-svg-icons";
import { ExerciseIcons } from "./Exercise";

export const StyledToolbar = styled.div<{
  mobile: boolean;
  desktop: boolean;
}>`
  position: absolute;
  background: #fff;
  width: 4.5rem;
  border: 1px solid ${colors.palette.light};
  background: #fff;
  margin: ${layouts.card.spacing} 0 0 ${layouts.card.spacing};
  z-index: 1;

  > * {
    height: 4.7rem;
    display: flex;
    align-items: center;
    justify-content: center;
    color: ${colors.palette.neutral};
    background: #fff;
    text-decoration: none;

    &:after {
      border-bottom: 1px solid #f1f1f1;
    }

    svg {
      width: 4.3rem;
      height: 1.6rem;
      transition: color 150ms;
    }

    span {
      display: none;
      flex-grow: 1;
      font-size: 1.4rem;
      color: ${colors.palette.neutralDarker};
    }

    &:hover {
      box-shadow: rgba(0, 0, 0, 0.2) 0px 0px 2px;
      width: 20rem;

      svg {
        color: ${colors.palette.neutralDarker};
      }

      span {
        display: block;
      }
    }
  }

  ${props => !props.desktop && css`
    ${breakpoints.desktop`
      display: none;
    `}
    ${breakpoints.tablet`
      display: none;
    `}
  `}

  ${props => props.mobile && breakpoints.mobile`
    display: block;
    position: relative;
    width: auto;
    display: flex;
    flex-direction: row;
    margin: 0;

    &:not(:last-child) {
      border-bottom: 0;
    }

    > * {
      flex-direction: column;
      flex-grow: 1;
      align-items: center;
      justify-content: center;
      padding: 1rem 1.8rem 0.6rem;
      height: auto;
      min-height: 4.8rem;

      &:hover {
        width: auto;
      }

      span {
        display: block;
        font-size: 1.4rem;
        margin-top: 0.2rem;
        color: ${colors.palette.neutral};
        text-align: center;
      }

      & + * {
        border-left: 1px solid ${colors.palette.light};
      }
    }
  `}
`;

export const ExerciseToolbar = ({ icons }: { icons?: ExerciseIcons }) => {
  if (!icons) {
    return null;
  }
  const settings = Object.values(icons);
  const mobile = settings.some(({ location }) => location?.toolbar?.mobile ?? true);
  const desktop = settings.some(({ location }) => location?.toolbar?.desktop ?? false);
  const topicUrl = icons.topic?.url;
  const errataUrl = icons.errata?.url;

  return <StyledToolbar {...{ mobile, desktop }}>
    {topicUrl ? <a href={topicUrl} target="_blank">
      <FontAwesomeIcon icon={faBookOpen}></FontAwesomeIcon>
      <span>View topic in textbook</span>
    </a> : null}
    {errataUrl ? <a href={errataUrl} target="_blank">
      <FontAwesomeIcon icon={faTriangleExclamation}></FontAwesomeIcon>
      <span>Suggest a correction</span>
    </a> : null}
  </StyledToolbar>
};
