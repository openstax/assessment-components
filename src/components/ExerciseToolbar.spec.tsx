import { ExerciseToolbar } from './ExerciseToolbar';
import renderer from 'react-test-renderer';

describe('ExerciseToolbar', () => {
  it('matches snapshot', () => {
    const instance = renderer.create(
      <ExerciseToolbar errataUrl='openstax.org' topicUrl='openstax.org' />
    );

    expect(instance.toJSON()).toMatchSnapshot();
  });

  it('conditionally renders icons', () => {
    expect(renderer.create(
      <ExerciseToolbar />
    ).toJSON()).toMatchInlineSnapshot(`null`);

    expect(renderer.create(
      <ExerciseToolbar errataUrl='openstax' />
    ).toJSON()).toMatchSnapshot();

    expect(renderer.create(
      <ExerciseToolbar topicUrl='openstax' />
    ).toJSON()).toMatchSnapshot();
  });
});
