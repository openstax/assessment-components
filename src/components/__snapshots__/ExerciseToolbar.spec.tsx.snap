// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseToolbar conditionally renders icons 2`] = `
<div
  className="sc-bczRLJ fXGBwS"
>
  <a
    href="openstax"
    target="_blank"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-circle-exclamation "
      data-icon="circle-exclamation"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 512c141.4 0 256-114.6 256-256S397.4 0 256 0S0 114.6 0 256S114.6 512 256 512zm0-384c13.3 0 24 10.7 24 24V264c0 13.3-10.7 24-24 24s-24-10.7-24-24V152c0-13.3 10.7-24 24-24zm32 224c0 17.7-14.3 32-32 32s-32-14.3-32-32s14.3-32 32-32s32 14.3 32 32z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <span>
      Suggest a correction
    </span>
  </a>
</div>
`;

exports[`ExerciseToolbar conditionally renders icons 3`] = `
<div
  className="sc-bczRLJ fXGBwS"
>
  <a
    href="openstax"
    target="_blank"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-book-open "
      data-icon="book-open"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 576 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M249.6 471.5c10.8 3.8 22.4-4.1 22.4-15.5V78.6c0-4.2-1.6-8.4-5-11C247.4 52 202.4 32 144 32C87.5 32 35.1 48.6 9 59.9c-5.6 2.4-9 8-9 14V454.1c0 11.9 12.8 20.2 24.1 16.5C55.6 460.1 105.5 448 144 448c33.9 0 79 14 105.6 23.5zm76.8 0C353 462 398.1 448 432 448c38.5 0 88.4 12.1 119.9 22.6c11.3 3.8 24.1-4.6 24.1-16.5V73.9c0-6.1-3.4-11.6-9-14C540.9 48.6 488.5 32 432 32c-58.4 0-103.4 20-123 35.6c-3.3 2.6-5 6.8-5 11V456c0 11.4 11.7 19.3 22.4 15.5z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <span>
      View topic in textbook
    </span>
  </a>
</div>
`;

exports[`ExerciseToolbar matches snapshot 1`] = `
<div
  className="sc-bczRLJ fXGBwS"
>
  <a
    href="openstax.org"
    target="_blank"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-book-open "
      data-icon="book-open"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 576 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M249.6 471.5c10.8 3.8 22.4-4.1 22.4-15.5V78.6c0-4.2-1.6-8.4-5-11C247.4 52 202.4 32 144 32C87.5 32 35.1 48.6 9 59.9c-5.6 2.4-9 8-9 14V454.1c0 11.9 12.8 20.2 24.1 16.5C55.6 460.1 105.5 448 144 448c33.9 0 79 14 105.6 23.5zm76.8 0C353 462 398.1 448 432 448c38.5 0 88.4 12.1 119.9 22.6c11.3 3.8 24.1-4.6 24.1-16.5V73.9c0-6.1-3.4-11.6-9-14C540.9 48.6 488.5 32 432 32c-58.4 0-103.4 20-123 35.6c-3.3 2.6-5 6.8-5 11V456c0 11.4 11.7 19.3 22.4 15.5z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <span>
      View topic in textbook
    </span>
  </a>
  <a
    href="openstax.org"
    target="_blank"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-circle-exclamation "
      data-icon="circle-exclamation"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 512c141.4 0 256-114.6 256-256S397.4 0 256 0S0 114.6 0 256S114.6 512 256 512zm0-384c13.3 0 24 10.7 24 24V264c0 13.3-10.7 24-24 24s-24-10.7-24-24V152c0-13.3 10.7-24 24-24zm32 224c0 17.7-14.3 32-32 32s-32-14.3-32-32s14.3-32 32-32s32 14.3 32 32z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <span>
      Suggest a correction
    </span>
  </a>
</div>
`;
